Update terraform docs (#596)
OctopusDeploy/docs: docs/deployment-examples/terraform-deployments/remote-state/index.md

---
title: Terraform Remote State
description: Using remote state with Octopus
---

> When running Terraform on a local PC, the state of the resources managed by Terraform is saved in a local file. This state is queried to learn which resources already exist in order to properly apply updates and destroy resources.

Reviewer: > the resources managed by Terraform is saved
I think it needs to be either "the resource is saved" or "the resources are saved"

> When Terraform is run by Octopus, this state file is not preserved between executions. This means that, for almost all practical applications of Terraform through Octopus, a remote backend must be configured. A remote backend allows this state information to be preserved between Terraform steps.

Reviewer 2: Slight tweak to change the order of the sentence and avoid the parenthetical phrase:
```suggestion
When Terraform is run by Octopus, this state file is not preserved between executions. This means a remote backend must be configured for almost all practical applications of Terraform through Octopus, allowing the state information to be preserved between Terraform steps.
```

Refer to the [Terraform documentation](https://www.terraform.io/docs/backends/index.html) for more information on configuring backends.

:::warning

> While neither Octopus nor Terraform will generate errors if a remote backend is not configured, most attempts to update or delete existing resources will not work as expected without a remote backend.

Reviewer 2: ```suggestion
Neither Octopus or Terraform will generate errors if a remote backend is not configured, but most attempts to update or delete existing resources will not work as expected without a remote backend.
```

:::

## Remote State Terraform Cloud

Using terraform enterprise for remote state requires a data source using referencing the `remote` backend

```
variable "token" {
  type = "string"
}

variable "organization" {
  type = "string"
}

variable "workspace" {
  type = "string"
}

data "terraform_remote_state" "state" {
	backend = "remote"
  	config = {
    	organization = "${var.organization}"
		workspaces = {
  			name = "${var.workspace}"
		}
		token =  "${var.token}"
  }
}
```

> Just like any other data source, it must first exist remotely. In order to achieve this, you must have an empty template such as above which contains only the data source in question. You should then run `terraform init` followed by

Reviewer 2: Minor tweaks to reduce the wordage:
```suggestion
As with any other data source, it must exist remotely first. To achieve this, you need an empty template as above which contains only the data source in question. You then need to run `terraform init` followed by
```

> `terraform plan` which will generate the empty state. The remote state can then be seeded using `terraform state push .\.terraform\terraform.tfstate`. This is necessary as including resources as part of the template will result in errors such as

Reviewer 2: ```suggestion
`terraform plan` to generate the empty state. The remote state can then be seeded using `terraform state push .\.terraform\terraform.tfstate`. This is necessary as including resources as part of the template will result in errors such as
```

`No stored state was found for the given workspace in the given backend.` as terraform tries to first read the remote state that doesn't exist.

## Enhanced Backends

> Terraform has the concept of enhanced backends which enable both storage and execution of operations such as plan and apply remotely. Octopus does not prevent you from using backends such as these, however the execution of actions remotely may not

Reviewer: ```suggestion
Terraform has the concept of enhanced backends which enable both storage and execution of remote operations such as plan and apply. Octopus does not prevent you from using backends such as these, however the execution of actions remotely may not
```
